File: backend/license-plate-app/api/models/region.py
from datetime import datetime
from enum import Enum
from typing import List, Optional, Union

from base.models import BaseModel
from base.schema import IDSchema, PaginationInfo
from pydantic import EmailStr, Field,AnyUrl
from utils.pyobjectid import ObjectId, PyObjectId
from .in_and_out_time import Type


class CameraRTSPModel(BaseModel):
    name: Optional[str]
    rtsp_url: Optional[AnyUrl]
    type: Optional[Type]

class RegionType(str,Enum):
    parking = 'parking'
    campus = 'campus'

class Coordinate(BaseModel):
    longitude: Optional[str]
    latitude: Optional[str]

class RegionModel(BaseModel):
    region: str = Field(...)
    type: RegionType = Field(default_factory=RegionType,alias='type')
    coordinate: Optional[Coordinate]
    cameras: Optional[List[CameraRTSPModel]] = []
    acceptance_roles: Optional[List[PyObjectId]] = []

    class Config:
        schema_extra = {
            "example": {
                "region": "str",
                "type": "str",
                "coordinate":{
                    "longitude":"str",
                    "latitude":"str"
                },
            }
        }

class UpdateRegionModel(BaseModel):
    region: Optional[str]
    type: Optional[RegionType]
    coordinate: Optional[Coordinate]
    cameras: Optional[List[CameraRTSPModel]] = []

    # class Config:
    #     schema_extra = {
    #         "example": {
    #             "region": "str",
    #             "type": "str",
    #             "coordinate":{
    #                 "longitude":"str",
    #                 "latitude":"str"
    #             },
    #         }
    #     }

class RegionModelOut(UpdateRegionModel,IDSchema):
    pass

class RegionModelListOut(PaginationInfo):
    list: Optional[List[RegionModelOut]]

class SearchRegionModel(BaseModel):
    id_region: Optional[PyObjectId]
    type: Optional[RegionType]

File: mobile/lib/core/models/Region.dart
class Region {
  String? sId;
  String? region;
  String? type;
  Coordinate? coordinate;
  List<Cameras>? cameras;

  Region({this.sId, this.region, this.type, this.coordinate, this.cameras});

  Region.fromJson(Map<String, dynamic> json) {
    sId = json['_id'];
    region = json['region'];
    type = json['type'];
    coordinate = json['coordinate'] != null
        ? new Coordinate.fromJson(json['coordinate'])
        : null;
    if (json['cameras'] != null) {
      cameras = <Cameras>[];
      json['cameras'].forEach((v) {
        cameras!.add(new Cameras.fromJson(v));
      });
    }
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['_id'] = this.sId;
    data['region'] = this.region;
    data['type'] = this.type;
    if (this.coordinate != null) {
      data['coordinate'] = this.coordinate!.toJson();
    }
    if (this.cameras != null) {
      data['cameras'] = this.cameras!.map((v) => v.toJson()).toList();
    }
    return data;
  }
}

class Coordinate {
  String? longitude;
  String? latitude;

  Coordinate({this.longitude, this.latitude});

  Coordinate.fromJson(Map<String, dynamic> json) {
    longitude = json['longitude'];
    latitude = json['latitude'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['longitude'] = this.longitude;
    data['latitude'] = this.latitude;
    return data;
  }
}

class Cameras {
  String? name;
  String? rtspUrl;
  String? type;

  Cameras({this.name, this.rtspUrl, this.type});

  Cameras.fromJson(Map<String, dynamic> json) {
    name = json['name'];
    rtspUrl = json['rtsp_url'];
    type = json['type'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['name'] = this.name;
    data['rtsp_url'] = this.rtspUrl;
    data['type'] = this.type;
    return data;
  }
}
